pkg/proctree/tree.go

```diff
@@ -0,0 +1,297 @@
+	// There should be no connections to and from the node after this is called.
+	DisconnectNode()
+}
+
+type ProcessTreeConfig struct {
+	// Max amount of processes nodes to allow in memory.
+	// Too small value here might result missing information in the tree and inconsistency.
+	// Default recommended value is 32768, as this is the max amount of PIDs in the system.
+	// You might even want more than 32768 to allow relevant information of dead processes to be
+	// available after their exit.
+	MaxProcesses int
+	// Max amount of threads nodes to allow in memory.
+	// Too small value here might result missing information in the tree and inconsistency.
+	// Default recommended value is 32768, as this is the max amount of TIDs in the system.
+	// You might even want more than 32768 to allow relevant information of dead threads to be
+	// available after their exit.
+	MaxThreads int
+	// Max size of cache for processes removing after exit - translate to the delay between process
+	// exit to its removal from the tree (if it is not interesting after its death).
+	// Too small value here might result processes missing from tree after their exit, before
+	// users were able to query the process information because of the delay between the feeding
+	// of the tree and the consumption of its information.
+	MaxCacheDelete int
+}
+
+// ProcessTree is a struct which follow the state of processes during runtime of a system.
+// The process tree is updated through Tracee's events, and is designed to overcome problems that
+// may arise because of events consumption (like handling lost events).
+// For more information on the logic, please go to the package documentation.
+type ProcessTree struct {
+	processes          types.RWMap[int, *processNode]
+	threads            types.RWMap[int, *threadNode]
+	processesGC        *lru.Cache[nodeUniqueId, *processNode]
+	threadsGC          *lru.Cache[nodeUniqueId, *threadNode]
+	deadProcessesCache *lru.Cache[int, bool]
+}
+
+func NewProcessTree(config ProcessTreeConfig) (*ProcessTree, error) {
+	tree := &ProcessTree{
+		processes: types.InitRWMap[int, *processNode](),
+		threads:   types.InitRWMap[int, *threadNode](),
+	}
+	cache, err := lru.NewWithEvict[int, bool](
+		config.MaxCacheDelete,
+		func(dpid int, _ bool) {
+			tree.removeProcessFromTree(dpid)
+		})
+	if err != nil {
+		return nil, err
+	}
+	processesGC, err := lru.NewWithEvict[nodeUniqueId, *processNode](
+		config.MaxProcesses,
+		func(id nodeUniqueId, node *processNode) {
+			err = tree.deleteProcess(node)
+			if err != nil {
+				logger.Errorw("delete process from tree", "err", err.Error())
+			}
+		},
+	)
+	if err != nil {
+		return nil, err
+	}
+	threadsGC, err := lru.NewWithEvict[nodeUniqueId, *threadNode](
+		config.MaxThreads,
+		func(id nodeUniqueId, node *threadNode) {
+			err = tree.deleteThread(node)
+			if err != nil {
+				logger.Errorw("delete thread from tree", "err", err.Error())
+			}
```

Comment:
Could you explain what each cache does and why they exist ? If this is explained somewhere else, then I believe it should be moved here IMO.